### DISCUSSION TITLE
Custom hook to extend TanStack Table with Excel-like features

### DISCUSSION THREAD

**Discussion by odanieloliveira** [TanStack/table#5613] [Q&A]
Hi everyone,
I've built a package that adds some Excel-like features to tanstack table such as cell selection (single and range), copy/paste, history of changes, cell editing. I've reproduced a working example of the package in [this sandbox](https://codesandbox.io/p/sandbox/datatable-nrlr2v?file=%2Fsrc%2Fcolumns.tsx%3A16%2C23&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clxm5fci000063b6krpmdz83o%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clxm5fci000023b6k820fcaxy%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clxm5fci000033b6kmztwv2i9%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clxm5fci000053b6kaamxali4%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clxm5fci000023b6k820fcaxy%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxm5fci000013b6ky04g2opi%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clxm6qoyo00023b6m7wsd93d3%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A133%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A133%252C%2522endColumn%2522%253A1%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252Fstrains.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clxm6tg8100023b6k6f08dn4q%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A16%252C%2522startColumn%2522%253A23%252C%2522endLineNumber%2522%253A16%252C%2522endColumn%2522%253A23%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252Fcolumns.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clxm5fci000023b6k820fcaxy%2522%252C%2522activeTabId%2522%253A%2522clxm6tg8100023b6k6f08dn4q%2522%257D%252C%2522clxm5fci000053b6kaamxali4%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxm5fci000043b6k9surgdhp%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clxm5fci000053b6kaamxali4%2522%252C%2522activeTabId%2522%253A%2522clxm5fci000043b6k9surgdhp%2522%257D%252C%2522clxm5fci000033b6kmztwv2i9%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clxm5fci000033b6kmztwv2i9%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D).

Even though this particular example works well, I have a more complex example that isn't fully working as expected but I can't reproduce in a sandbox due to dependencies, I've added it under a "broken code" folder in the sandbox. More specifically, when editing cells, the new values don't seem to be reflected when logging `table.getRowModel().rows[rowId].getValue(columnId)` but they do show up when logging `table.getRowModel().rows[rowId].original[columnId]` (which I didn't expect since this is supposed to store the original values). The paste functionality is also not working in this particular example and I have no clue why. 

I understand it's hard to help without the full code to reproduce but I'm hopeful I'm just doing something obviously wrong to more experienced members of the community and also curious if you have any general feedback on my code and see obvious improvements to what I've done here.

**Comment by wi-ski** [comment]
This is awesome - just wanted you to know. Great work.

**Comment by mmospanenko** [comment]
Awesome, thank you!

**Comment by Radinax** [comment]
YOU'RE THE GOAT! This is extremely good, well done.
